proxy/serve.go

```diff
@@ -177,3 +197,14 @@ func serve(ln net.Listener, srv Server) error {
 	}
 	return err
 }
+
+func removeServer(srvs []Server, s Server) []Server {
+	for i, srv := range srvs {
+		if s == srv {
```

Comment:
I'm a bit concerned this is going to be brittle.  A better, if more invasive solution might be to tweak the interface for Server here so that we can add a method to return the listen address, and then compare the returns from the Address() methods.  We need to add this method to the existing implementations of course.

Reply:
on second thought, I'm thinking that instead of changing the interface to have an Address() method, maybe just extend the use of your map to include all servers, i.e., change servers []Server to servers map[string]Server , and the key is the listen address .  Thoughts?

Reply:
If we change `servers` from slice to map what about `ListenAndServeHTTPSTCPSNI()` which passes `nil` as a listener? Should we pass the result of `tls.NewListener(httpsListener, cfg)` (line no. 139) anyway just to extract address and use it as a key in the `servers` map?

Reply:
That's certainly an option.  In that section, httpsListener, the TLS wrapped version of httpsListener that's passed into ServeLater() and tps are net.Listeners which will return the same Addr() .  Another option is to avoid using the Addr() method of net.Listener and just add a string parameter to the serve() method.